lib/platform-stats.js

```diff
@@ -99,30 +93,153 @@ export const updateDailyStationStats = async (
 
 /**
  * @param {import('pg').Client} pgClient
- * @param {Set<string>} participants
+ * @param {number[]} participantIds
  */
-export const updateDailyParticipants = async (pgClient, participants) => {
-  debug('Updating daily participants, count=%s', participants.size)
-  const ids = await mapParticipantsToIds(pgClient, participants)
+export const updateDailyParticipants = async (pgClient, participantIds) => {
+  debug('Updating daily participants, count=%s', participantIds.length)
   await pgClient.query(`
     INSERT INTO daily_participants (day, participant_id)
     SELECT now() as day, UNNEST($1::INT[]) AS participant_id
     ON CONFLICT DO NOTHING
   `, [
-    ids
+    participantIds
+  ])
+}
+
+/**
+ * @param {import('pg').Client} pgClient
+ * @param {import('./preprocess.js').Measurement[]} honestMeasurements
+ * @param {import('./preprocess.js').Measurement[]} allMeasurements
+ * @param {Map<string, number>} participantsMap
+ * @param {object} options
+ * @param {Date} [options.day]
+ */
+export const updateStationsAndParticipants = async (
+  pgClient,
+  honestMeasurements,
+  allMeasurements,
+  participantsMap,
+  { day = new Date() } = {}
+) => {
+  /** @type {Map<number, Map<string, {accepted: number, total: number}>>} */
+  const participantsStationsStats = new Map()
+
+  /** @type {Map<number, Set<string>>} */
+  const subnets = new Map()
+
+  /** @type {(m: import('./preprocess.js').Measurement, participantId: number, countType: 'accepted' | 'total') => void} */
+  const updateParticipantsStationsStats = (m, participantId, countType) => {
+    if (participantId == null) return
+
+    let participant = participantsStationsStats.get(participantId)
+    if (!participant) {
+      participant = new Map()
+      participantsStationsStats.set(participantId, participant)
+    }
+
+    let stationStats = participant.get(m.stationId)
+    if (!stationStats) {
+      stationStats = { accepted: 0, total: 0 }
+      participant.set(m.stationId, stationStats)
+    }
+
+    stationStats[countType]++
+  }
+
+  for (const m of honestMeasurements) {
+    if (m.stationId == null) continue
+    const participantId = participantsMap.get(m.participantAddress)
+    if (participantId == null) continue
+
+    updateParticipantsStationsStats(m, participantId, 'accepted')
+
+    const participantSubnets = subnets.get(participantId) ?? new Set()
+    participantSubnets.add(m.inet_group)
+    subnets.set(participantId, participantSubnets) // in case new set was created
+  }
+
+  for (const m of allMeasurements) {
+    if (m.stationId == null) continue
+    const participantId = participantsMap.get(m.participantAddress)
+    if (participantId == null) continue
+
+    updateParticipantsStationsStats(m, participantId, 'total')
+  }
```

Comment:
Do we want to also add to the participantSubnets for allMeasurements?

```
    const participantSubnets = subnets.get(participantId) ?? new Set()
    participantSubnets.add(m.inet_group)
    subnets.set(participantId, participantSubnets) // in case new set was created
```

We were already not doing this, but I don't recall if there was a specific reason.

Reply:
> Do we want to also add to the participantSubnets for allMeasurements?

Yes. We should look at `allMeasurements` when counting anything related to participants - participant addresses, IP subnets, and station IDs too.

Could you please check if we collect station IDs from all measurements, not only the accepted ones?

Reply:
Gotcha, adding that in now.

Can you elaborate on what you mean by "collect station IDs"? We call `updateParticipantsStationsStats` in the code attached to this comment thread for allMeasurements to increment the total, is that what you mean?

Reply:
I was wondering whether the code inserting data to `daily_stations` would include Station IDs whose measurements we rejected (Station IDs with no accepted measurements).

I checked the source code of `updateDailyStationStats` and verified that we include those Stations IDs.

All is good.

https://github.com/filecoin-station/spark-evaluate/blob/ae885a9dcac95a9c3265634511eae5779e7370a5/lib/platform-stats.js#L25C14-L98